app/models/sms/adapters/frontline_cloud_adapter.rb

```diff
@@ -0,0 +1,50 @@
+class Sms::Adapters::FrontlineCloudAdapter < Sms::Adapters::Adapter
+
+
+  def self.recognize_receive_request?(request)
+    %w(from body sent_at frontlinecloud) - request.params.keys == []
+  end
+
+  def self.can_deliver?
+    true
+  end
+
+  def reply_style
+    :via_adapter
+  end
+
+  def deliver(message)
+    prepare_message_for_delivery(message)
+
+    # build the request payload
+    recipients = message.recipient_numbers.map { |number| { "type" => "mobile", "value" => number } }
+
+    payload = {
+      "apiKey" => configatron.frontlinecloud_api_key,
+      "payload" => {
+        "message" => message.body,
+        "recipients" => recipients
+      },
+    }
+    uri = URI("https://cloud.frontlinesms.com/api/1/webhook")
+
+    response = send_request(uri, :post, payload)
+  end
+
+  def receive(request)
+    params = request.params
+    Sms::Incoming.new(
+      from: params["from"],
+      to: nil, # Frontline doesn't provide this.
+      body: params["body"],
+      sent_at: convert_time(params["sent_at"]),
+      adapter_name: service_name)
+  end
+
+  private
+
+  def convert_time(timestamp)
+    timestamp = timestamp.to_i
+    datetime = Time.at(timestamp).to_datetime
```

Comment:
It does, DateTimes and Times mostly quack the same way, but I think most of the other time objects being passed around are DateTimes and I wanted to make sure they were getting the objects they were expecting.